common/component_interface_utils/README.md

```diff
@@ -4,29 +4,117 @@
 
 This is a utility package that provides the following features:
 
+- Instantiation of the wrapper class
 - Logging for service and client
+- Service exception for response
+- Macros
+- Relays for topic and service
 
-## Usage
+## Design
 
-1. This package requires interface information in this format.
+This package provides the wrappers for the interface classes of rclcpp.
+The wrappers limit the usage of the original class to enforce the processing recommended by the component interface.
+Do not inherit the class of rclcpp, and forward or wrap the member function that is allowed to be used.
 
-   ```cpp
-   struct SampleService
-   {
-     using Service = sample_msgs::srv::ServiceType;
-     static constexpr char name[] = "/sample/service";
-   };
-   ```
+## Instantiation of the wrapper class
 
-2. Create a wrapper using the above definition as follows.
+The wrapper class requires interface information in this format.
 
-   ```cpp
-   component_interface_utils::Service<SampleService>::SharedPtr srv_;
-   srv_ = component_interface_utils::create_service<SampleService>(node, ...);
-   ```
+```cpp
+struct SampleService
+{
+  using Service = sample_msgs::srv::ServiceType;
+  static constexpr char name[] = "/sample/service";
+};
 
-## Design
+struct SampleMessage
+{
+  using Message = sample_msgs::msg::MessageType;
+  static constexpr char name[] = "/sample/message";
+  static constexpr size_t depth = 3;
+  static constexpr auto reliability = RMW_QOS_POLICY_RELIABILITY_RELIABLE;
+  static constexpr auto durability = RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL;
+};
+```
 
-This package provides the wrappers for the interface classes of rclcpp.
-The wrappers limit the usage of the original class to enforce the processing recommended by the component interface.
-Do not inherit the class of rclcpp, and forward or wrap the member function that is allowed to be used.
+Create the wrapper using the above definition as follows.
+
+```cpp
+// header file
+component_interface_utils::Service<SampleService>::SharedPtr srv_;
+component_interface_utils::Client<SampleService>::SharedPtr cli_;
+component_interface_utils::Publisher<SampleMessage>::SharedPtr pub_;
+component_interface_utils::Subscription<SampleMessage>::SharedPtr sub_;
+
+// source file
+const auto node = component_interface_utils::NodeAdaptor(this);
+node.init_srv(srv_, callback);
+node.init_cli(srv_);
+node.init_pub(srv_);
+node.init_sub(srv_, callback);
+```
+
+## Logging for service and client
+
+If the wrapper class is used, logging is automatically enabled. The log level is `RCLCPP_INFO`.
+
+## Service exception for response
+
+If the wrapper class is used, throwing `ServiceException` will automatically catch it and set it to response status.
+This is useful when returning an error from a function called from the service callback.
+
+```cpp
+void ServiceCallback(Request req, Response res)
```

Comment:
There is no particular reason. Is this rule only for global functions? Or does it apply to member functions?

Reply:
I believe it also applies to member functions.

@xmfcx Do you have any comments on the function naming convention?

Reply:
@kenji-miyake I used to use CamelCase for functions because like it says [in ros2 guidelines](https://docs.ros.org/en/foxy/The-ROS2-Project/Contributing/Code-Style-Language-Versions.html#function-and-method-naming) that it was the recommended way from google-cpp guidelines.

But to be more ros-like, I've changed my function naming convention to snake_case. It looked weird at first but now I got used to it. And I think entire universe code base uses snake_case, so let's go with the snake_case 🤷‍♂️

Reply:
Renamed https://github.com/autowarefoundation/autoware.universe/pull/1429/commits/34461b5d2700c5fb49dd49a7b1b6cd6666caa6fd.